### ISSUE TITLE
Measure range

### ISSUE THREAD

**Issue by milorenzo** [javier2491/FaceQnet#3]
Hello,

I've been trying your code with several experiments. 

I've modified the input so I can predict the quality scores live from a webcam, but it seems very difficult to reach scores outside the 0.6-0.75 range. Fig 5 from the paper shows images retrieving 0.1, 0.25... but I can't reproduce such scores. In fact, I've encountered some weird results like this one:

![image](https://user-images.githubusercontent.com/49560701/56040728-56103d00-5d37-11e9-9965-79401df9e8cd.png)

Am I missing something?

Best regards.

**Comment by forseeking** [comment]
@uam-biometrics Are these expected output of the sample image? It seems that the variation of the score is small.
./Samples_cropped/Picture 11.png;[0.73921335]
./Samples_cropped/Picture 12.png;[0.72319686]
./Samples_cropped/Picture 14.png;[0.70089]
./Samples_cropped/Picture 16.png;[0.60240257]
./Samples_cropped/Picture 17.png;[0.6484765]
./Samples_cropped/Picture 18.png;[0.68437815]
./Samples_cropped/Picture 19.png;[0.620492]
./Samples_cropped/Picture 20.png;[0.631745]
./Samples_cropped/Picture 21.png;[0.6528319]
./Samples_cropped/Picture 22.png;[0.626657]
./Samples_cropped/Picture 23.png;[0.6322764]
./Samples_cropped/Picture 24.png;[0.59827983]
./Samples_cropped/Picture 25.png;[0.58728033]
./Samples_cropped/Picture 26.png;[0.5906831]
./Samples_cropped/Picture 27.png;[0.6237298]
./Samples_cropped/Picture 28.png;[0.61860085]
./Samples_cropped/Picture 29.png;[0.67180717]
./Samples_cropped/Picture 30.png;[0.6043032]
./Samples_cropped/Picture 32.png;[0.590117]
./Samples_cropped/Picture 35.png;[0.6397052]
./Samples_cropped/Picture 36.png;[0.642903]
./Samples_cropped/Picture10.png;[0.74740803]

**Comment by javier-hernandezo** [comment]
> Hello,
> 
> I've been trying your code with several experiments.
> 
> I've modified the input so I can predict the quality scores live from a webcam, but it seems very difficult to reach scores outside the 0.6-0.75 range. Fig 5 from the paper shows images retrieving 0.1, 0.25... but I can't reproduce such scores. In fact, I've encountered some weird results like this one:
> 
> ![image](https://user-images.githubusercontent.com/49560701/56040728-56103d00-5d37-11e9-9965-79401df9e8cd.png)
> 
> Am I missing something?
> 
> Best regards.

Hello,

The script provided here adds a tanh normalization step after getting the outputs from FaceQnet, in order to constraint the output to the [0,1] range. This normalization step could be dependent of the type of input data, so I think you could try to remove that step from the script. This should give you more spread scores. 

Regarding your second question, Did you feed FaceQnet with the whole image or only with the region into the green rectangle? If you send me the image for replicating the score I could give you more information about what is happening.

Regards

**Comment by javier-hernandezo** [comment]
> @uam-biometrics Are these expected output of the sample image? It seems that the variation of the score is small.
> ./Samples_cropped/Picture 11.png;[0.73921335]
> ./Samples_cropped/Picture 12.png;[0.72319686]
> ./Samples_cropped/Picture 14.png;[0.70089]
> ./Samples_cropped/Picture 16.png;[0.60240257]
> ./Samples_cropped/Picture 17.png;[0.6484765]
> ./Samples_cropped/Picture 18.png;[0.68437815]
> ./Samples_cropped/Picture 19.png;[0.620492]
> ./Samples_cropped/Picture 20.png;[0.631745]
> ./Samples_cropped/Picture 21.png;[0.6528319]
> ./Samples_cropped/Picture 22.png;[0.626657]
> ./Samples_cropped/Picture 23.png;[0.6322764]
> ./Samples_cropped/Picture 24.png;[0.59827983]
> ./Samples_cropped/Picture 25.png;[0.58728033]
> ./Samples_cropped/Picture 26.png;[0.5906831]
> ./Samples_cropped/Picture 27.png;[0.6237298]
> ./Samples_cropped/Picture 28.png;[0.61860085]
> ./Samples_cropped/Picture 29.png;[0.67180717]
> ./Samples_cropped/Picture 30.png;[0.6043032]
> ./Samples_cropped/Picture 32.png;[0.590117]
> ./Samples_cropped/Picture 35.png;[0.6397052]
> ./Samples_cropped/Picture 36.png;[0.642903]
> ./Samples_cropped/Picture10.png;[0.74740803]

Hello,

Yes, those are the expected output scores. The images provided for testing have different levels of quality but all of them present factors such as good resolution, illumination, etc that make their scores to be in the [0.5, 0.8] range. 
However, the script provided here adds a tanh normalization step after getting the outputs from FaceQnet, in order to constraint the output to the [0,1] range. You could try to remove that step from the script. That should give you more spread scores.

Regards

**Comment by forseeking** [comment]
> > @uam-biometrics Are these expected output of the sample image? It seems that the variation of the score is small.
> > ./Samples_cropped/Picture 11.png;[0.73921335]
> > ./Samples_cropped/Picture 12.png;[0.72319686]
> > ./Samples_cropped/Picture 14.png;[0.70089]
> > ./Samples_cropped/Picture 16.png;[0.60240257]
> > ./Samples_cropped/Picture 17.png;[0.6484765]
> > ./Samples_cropped/Picture 18.png;[0.68437815]
> > ./Samples_cropped/Picture 19.png;[0.620492]
> > ./Samples_cropped/Picture 20.png;[0.631745]
> > ./Samples_cropped/Picture 21.png;[0.6528319]
> > ./Samples_cropped/Picture 22.png;[0.626657]
> > ./Samples_cropped/Picture 23.png;[0.6322764]
> > ./Samples_cropped/Picture 24.png;[0.59827983]
> > ./Samples_cropped/Picture 25.png;[0.58728033]
> > ./Samples_cropped/Picture 26.png;[0.5906831]
> > ./Samples_cropped/Picture 27.png;[0.6237298]
> > ./Samples_cropped/Picture 28.png;[0.61860085]
> > ./Samples_cropped/Picture 29.png;[0.67180717]
> > ./Samples_cropped/Picture 30.png;[0.6043032]
> > ./Samples_cropped/Picture 32.png;[0.590117]
> > ./Samples_cropped/Picture 35.png;[0.6397052]
> > ./Samples_cropped/Picture 36.png;[0.642903]
> > ./Samples_cropped/Picture10.png;[0.74740803]
> 
> Hello,
> 
> Yes, those are the expected output scores. The images provided for testing have different levels of quality but all of them present factors such as good resolution, illumination, etc that make their scores to be in the [0.5, 0.8] range.
> However, the script provided here adds a tanh normalization step after getting the outputs from FaceQnet, in order to constraint the output to the [0,1] range. You could try to remove that step from the script. That should give you more spread scores.
> 
> Regards

I screen copied the image  in the paper which has a score of 0.1 but the code gives the score of 0.7994855, did I do something wrong ?
![Uploading 20190416101511.png…]()

**Comment by milorenzo** [comment]
> Hello,
> 
> The script provided here adds a tanh normalization step after getting the outputs from FaceQnet, in order to constraint the output to the [0,1] range. This normalization step could be dependent of the type of input data, so I think you could try to remove that step from the script. This should give you more spread scores.
> 
> Regarding your second question, Did you feed FaceQnet with the whole image or only with the region into the green rectangle? If you send me the image for replicating the score I could give you more information about what is happening.
> 
> Regards

Hello,

I've removed the tanh normalization but the relative behaviour seems the same. It just added an offset: now values are spread in the 0.3-0.5 range. Here are some other weird results:

![image](https://user-images.githubusercontent.com/49560701/56198212-c3bfb000-603a-11e9-9262-cde1e1aaca05.png) 
Score: [0.26320124]

![image](https://user-images.githubusercontent.com/49560701/56198290-e94cb980-603a-11e9-8446-edace321f052.png)
Score: [0.312468]

I am feeding FaceQNet with a square crop of the green rectangle region showed before, which is taken from the MTCNN face detector, as you suggested.

**Comment by milorenzo** [comment]
Here is my code for predicting the FaceQNet scores from a webcam.

```python
import cv2
import numpy as np

from keras.models import load_model
from mtcnn.mtcnn_face_detector import MtcnnFaceDetector

def main():
    # Loading the pretrained model
    print('Loading FaceQnet model...')
    q_model = load_model('FaceQnet.h5')

    print('Loading mtcnn...')
    mtcnn_face_detector = MtcnnFaceDetector(max_image_size_lowest_side=360)

    cap = cv2.VideoCapture(-1)
    if not cap:
        raise RuntimeError('OpenCV video capture failed')
    cap.set(cv2.CAP_PROP_FRAME_WIDTH, 1280)
    cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 720)
    while cap.isOpened():
        success, frame = cap.read()
        if success:
            frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            bboxes, landmarks, scores, _ = mtcnn_face_detector.detect(frame_rgb)
            if len(bboxes) > 0:
                faces = mtcnn_face_detector.crop_faces(frame, bboxes, square=True)

                test_data = []
                for face in faces:
                    face = cv2.resize(face, (224, 224), interpolation=cv2.INTER_AREA)
                    face_data = np.array(face, copy=False, dtype=np.float32)
                    test_data.append(face_data)
                q_scores = get_quality_score(q_model, test_data, batch_size=len(faces))

                for i, bbox in enumerate(bboxes):
                    face_top, face_left, face_bottom, face_right = bbox
                    pt1 = (int(face_left), int(face_top))
                    pt2 = (int(face_right), int(face_bottom))
                    cv2.rectangle(frame, pt1, pt2, (0, 255, 0), 2)
                    cv2.putText(frame, '{0:.2f}'.format(q_scores[i][0]), (pt2[0]-50, pt2[1]+25),
                                cv2.FONT_HERSHEY_SIMPLEX, 0.8, (0, 0, 255), 3)
            else:
                cv2.putText(frame, "Face not detected", (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 0.8, (255, 0, 0), 3)
        else:
            print('Cant retrieve frame')
        # Display the resulting frame
        cv2.imshow('FaceQnet demo', frame)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    cap.release()
    cv2.destroyAllWindows()

def get_quality_score(q_model, test_data, batch_size=1):
    # Extract quality scores for the samples
    m = 0.7
    s = 0.5
    score = q_model.predict(np.stack(test_data, axis=0), batch_size=batch_size, verbose=1)
    # score = 0.5 * np.tanh(((score - m) / s) + 1)
    predictions = -score + 1  # Convertimos el score de distancia en similitud
    predictions = np.clip(predictions, 0.0, 1.0)
    return predictions

if __name__ == "__main__":
    main()
```

**Comment by radao** [comment]
Also having difficulty interpreting the output scores. If you could clarify this point, that would be appreciated.

**Comment by javier-hernandezo** [comment]
I have read all your comments, and I have been checking the script and the model.
I want to apologize, the script provided was meant to be used with other model.
I have updated both the model and the script, please redownload them and try your experiments again.
The predictions should be more accurate now. 

Thank you everyone for your comments.

![bradpitt](https://user-images.githubusercontent.com/40597519/56254980-fc748d80-60fd-11e9-902f-e9e024957eaf.png)

score = [0.5185597]

![jonsnow](https://user-images.githubusercontent.com/40597519/56254992-0b5b4000-60fe-11e9-92fb-21a931360a3b.png)

 score = [0.5391177]

![monja](https://user-images.githubusercontent.com/40597519/56255004-17df9880-60fe-11e9-93ad-10ca416f8c17.png)

score = [0.2676753]

**Comment by milorenzo** [comment]
The behaviour now seems more accurate to the "perceptual" quality.

I just have a final question, when predicting the input image, which is the expected color space at the input of the network, BGR (returned by cv2.imread) or RGB?

Thank you very much @uam-biometrics

**Comment by radao** [comment]
Thanks @uam-biometrics! Definitely looking better but still have some questions regarding interpreting the scores in certain contexts.

From the paper, you used the ICAO Compliance framework to generate ground truth quality measures. These scores act as a proxy for 30 indicators the framework defines:

![Screenshot from 2019-04-17 11-42-22](https://user-images.githubusercontent.com/2007531/56278093-e4a10780-6105-11e9-96d5-5e65bb54e849.png)

If this is true, we'd expect the quality scores from the model to have some correspondence to these indicators. For example, head poses with pitch/yaw/roll greater than 8 degrees should be penalized, and I generally do see this reflected in the scores. However, other indicators that should be very easy for the model to pick up on, such as wearing sunglasses (indicator 26 - frame covering eyes), don't seem to influence the score much. Could this be because the training dataset did not include examples of these indicators? If you could offer some insights into how we can interpret the scores in context of the ICAO Compliance framework, as this how the ground truth is generated, that would be very valuable.

**Comment by javier-hernandezo** [comment]
> The behaviour now seems more accurate to the "perceptual" quality.
> 
> I just have a final question, when predicting the input image, which is the expected color space at the input of the network, BGR (returned by cv2.imread) or RGB?
> 
> Thank you very much @uam-biometrics

You are welcome! Answering your last question, the CNN is feed with the images just as they are returned by cv2.imread, that is BGR channels.

Regards

**Comment by javier-hernandezo** [comment]
> Thanks @uam-biometrics! Definitely looking better but still have some questions regarding interpreting the scores in certain contexts.
> 
> From the paper, you used the ICAO Compliance framework to generate ground truth quality measures. These scores act as a proxy for 30 indicators the framework defines:
> 
> ![Screenshot from 2019-04-17 11-42-22](https://user-images.githubusercontent.com/2007531/56278093-e4a10780-6105-11e9-96d5-5e65bb54e849.png)
> 
> If this is true, we'd expect the quality scores from the model to have some correspondence to these indicators. For example, head poses with pitch/yaw/roll greater than 8 degrees should be penalized, and I generally do see this reflected in the scores. However, other indicators that should be very easy for the model to pick up on, such as wearing sunglasses (indicator 26 - frame covering eyes), don't seem to influence the score much. Could this be because the training dataset did not include examples of these indicators? If you could offer some insights into how we can interpret the scores in context of the ICAO Compliance framework, as this how the ground truth is generated, that would be very valuable.

Hello again @radao,

Your question is an interesting one. Yes, the ICAO compliance framework consists of 30 different tests. However, we did not use all of them for obtaining the groundtruth for training FaceQnet, just a small selection. 
Due to space limitations in the paper we were not able of explaining which specific tests we used for it.
I do not have the code here to check it, but some tests we definitely used are:  blur, too dark/light, pixelation, varied background, roll/pitch/yaw, shadows, hat/cap, and glasses. 
As you mentioned, effectively not all these categories are fully represented in the training set, so maybe the results are not as accurate as desirable in some of them (e.g. glasses).
We are currently performing new experiments with more training data and a smarter selection and combination of the ICAO test in order to get a better quality groundtruth. We will published it when it is finished.

Regards.

**Comment by radao** [comment]
Thanks for the info @uam-biometrics. Indeed there are a lot of modifications that can be made depending on the desired application of the quality score.

It would be very interesting to have access to the training code to do research in this area. Do you plan on releasing it?

**Comment by javier-hernandezo** [comment]
@radao We do not have plans of releasing the source code at least in a close future, as we are still using it for new experiments. 
However, we plan to release new improved versions of FaceQnet if we achieve a significant improvements in its performance. We also plan to publish the quality scores obtained when processing some famous databases like LFW (Labeled Faces in the Wild).
Regards
